feat: add support for custom cache stores (`options.read/options.write`)

Mordred opened the pull request:
<!--
1. [Read and sign the CLA](https://cla.js.foundation/webpack/webpack.js.org). This needs to be done only once. PRs that haven't signed it won't be accepted.
2. Check out the [development guide](https://webpack.js.org/development/) for the API and development guidelines.
3. Read through the PR diff carefully as sometimes this can reveal issues. The work will be reviewed, but this can save some effort.
-->

Hi, I want to change how cache data are stored in my project. In my case I want to use Redis server like this:

```js
const redis = require('redis');

// ...
// ... connect to client
// ...

const BUILD_CACHE_TIMEOUT = 24 * 3600; // 1 day

function redisKey(filename) {
    return `build:cache:${filename}`;
}

function readCache(filename, callback) {
      client.get(redisKey(filename), (err, result) => {
          if (err) {
              return callback(err);
          }

          if (!result) {
              return callback(new Error(`Key ${redisKey(filename)} not found`));
          }

          callback(null, result);
    });
}

function writeCache(filename, content, callback) {
      client.set(redisKey(filename), content, 'EX', BUILD_CACHE_TIMEOUT, callback);
}
```

Then in my webpack config:
```js
{
  loader: 'cache-loader',
  options: {
    readCache,
    writeCache,
  },
},
```

For that I needed `writeCache` and `readCache` hooks.

michael-ciniawsky commented on README.md:
```diff
@@ -49,6 +49,8 @@ module.exports = {
 |:--:|:--:|:-----:|:----------|
 |**`cacheDirectory`**|`{String}`|`path.resolve('.cache-loader')`|Provide a cache directory where cache items should be stored|
 |**`cacheIdentifier`**|`{String}`|`cache-loader:{version} {process.env.NODE_ENV}`|Provide an invalidation identifier which is used to generate the hashes. You can use it for extra dependencies of loaders.|
+|**`writeCache`**|`{Function(cacheFilename, content, callback) -> {void}}`|`undefined`|Allows you to override default write cache data to file (e.g. Redis, memcached).|
```
Just `writer` (`options.writer`) ?

michael-ciniawsky commented on README.md:
```diff
@@ -49,6 +49,8 @@ module.exports = {
 |**`cacheDirectory`**|`{String}`|`path.resolve('.cache-loader')`|Provide a cache directory where cache items should be stored|
 |**`cacheIdentifier`**|`{String}`|`cache-loader:{version} {process.env.NODE_ENV}`|Provide an invalidation identifier which is used to generate the hashes. You can use it for extra dependencies of loaders.|
+|**`writeCache`**|`{Function(cacheFilename, content, callback) -> {void}}`|`undefined`|Allows you to override default write cache data to file (e.g. Redis, memcached).|
+|**`readCache`**|`{Function(cacheFilename, callback) -> {void}}`|`undefined`|Allows you to override default read cache data from file.|
```
Just `reader` (`options.reader`) ?

michael-ciniawsky commented on README.md:
```diff
@@ -49,6 +49,8 @@ module.exports = {
+|**`writeCache`**|`{Function(cacheFilename, content, callback) -> {void}}`|`undefined`|Allows you to override default write cache data to file (e.g. Redis, memcached).|
+|**`readCache`**|`{Function(cacheFilename, callback) -> {void}}`|`undefined`|Allows you to override default read cache data from file.|
 
 <h2 align="center">Examples</h2>
```
Please add an example aswell

Mordred commented:
@michael-ciniawsky `fs` is faster than database, but we are building application for production in the docker. Saving cached data to filesystem is not an option because mounting volumes during `docker build` is not possible.

I have changed option names and added example from my first post.

michael-ciniawsky commented on README.md:
````diff
@@ -95,6 +97,62 @@ module.exports = {
 }
 ```
 
+### `Write cache data to redis`
````
Could you please try to make this example a bit more generic? e.g `Database Intergation` something like that :)

sokra commented:
cache-loader should no longer talk about filenames. Filenames should only occur in the default read/write/key implementation.

Mordred commented:
@sokra refactored

michael-ciniawsky commented on README.md:
````diff
@@ -95,6 +98,82 @@ module.exports = {
 }
 ```
 
+### `Example of Database Integration`
````
`### Database Intergration`

michael-ciniawsky commented on src/index.js:
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
 const defaultCacheDirectory = path.resolve('.cache-loader');
 const ENV = process.env.NODE_ENV || 'development';
 
+const defaultOptions = {
```
`defaultOptions` => `defaults`

michael-ciniawsky commented on src/index.js:
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
 const defaultCacheDirectory = path.resolve('.cache-loader');
```
```diff
- const defaultCacheDirectory = path.resolve('.cache-loader');
```

michael-ciniawsky commented on src/index.js:
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
 const ENV = process.env.NODE_ENV || 'development';
 
+const defaultOptions = {
+  cacheDirectory: defaultCacheDirectory,
```
```
cacheDirectory: path.resolve('.cache-loader'),
```

michael-ciniawsky commented on src/index.js:
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
+};
+
 function loader(...args) {
+  const loaderOptions = loaderUtils.getOptions(this) || {};
```
```diff
- const loaderOptions = loaderUtils.getOptions(this) || {};
```

michael-ciniawsky commented on src/index.js:
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
+
 function loader(...args) {
+  const loaderOptions = loaderUtils.getOptions(this) || {};
+  const options = Object.assign({}, defaultOptions, loaderOptions);
```
```js
const options = Object.assign({}, defaults, loaderUtils.getOptions(this))
```

michael-ciniawsky commented on src/index.js:
```diff
@@ -35,62 +46,33 @@ function loader(...args) {
-  data.cacheFile = cacheFile;
-  fs.readFile(cacheFile, 'utf-8', (readFileErr, content) => {
-    if (readFileErr) {
+  data.cacheKey = cacheKey;
```
`\n data.cacheKey = cacheKey(options, remainingRequest) \n`

michael-ciniawsky commented on src/index.js:
```diff
@@ -123,4 +105,45 @@ function digest(str) {
   return crypto.createHash('md5').update(str).digest('hex');
 }
 
+const directoriesExists = {};
```
`const directories = new Set();\n`

michael-ciniawsky commented on src/index.js:
```diff
@@ -123,4 +105,45 @@ function digest(str) {
+const directoriesExists = {};
+function write(cacheKey, data, callback) {
+  const dirname = path.dirname(cacheKey);
+  const content = JSON.stringify(data);
```
`\n`

michael-ciniawsky commented on src/index.js:
```diff
@@ -123,4 +105,45 @@ function digest(str) {
+function write(cacheKey, data, callback) {
+  const dirname = path.dirname(cacheKey);
+  const content = JSON.stringify(data);
+  if (directoriesExists[dirname]) {
```
`if (directories.has(dirname)) {`

michael-ciniawsky commented on src/index.js:
```diff
@@ -123,4 +105,45 @@ function digest(str) {
+        callback(mkdirErr);
+        return;
+      }
+      directoriesExists[dirname] = true;
```
`\ndirectories.add(dirname)`

michael-ciniawsky commented on src/index.js:
```diff
@@ -123,4 +105,45 @@ function digest(str) {
+  });
+}
+
+function makeKey(options, request) {
```
`function cacheKey(options, request) {`

michael-ciniawsky commented on src/index.js:
```diff
@@ -123,4 +105,45 @@ function digest(str) {
+
+function makeKey(options, request) {
+  const { cacheIdentifier, cacheDirectory } = options;
+  const hash = digest(`${cacheIdentifier}\n${request}`);
```
`\n`

michael-ciniawsky commented on src/index.js:
```diff
@@ -35,62 +46,33 @@ function loader(...args) {
   const options = Object.assign({}, defaultOptions, loaderOptions);
-  const { cacheIdentifier, cacheDirectory } = options;
+  const { read: readFn, makeKey: makeKeyFn } = options;
   const data = dataInput;
```
Please move this to L67

Crazymax11 commented:
Hello
We also tried to use cache loader to build application in docker and met two problems:
* Cache loader uses mtime to check if dependency changed, but git doesn't store files mtime, so after git clone we always have new mtime and no cache used.
* It seems that cache loader doesn't use content of resource to check if resources changed, cache loader just checks if dependencies changed.

It would be nice to know if module changed by checking hash content instead of mtime. It would very nice to provide an api to set `checkIfChanged` function as option.

Can it be done in this PR?

Mordred commented:
@michael-ciniawsky `makeKey` -> `cacheKey`

@Crazymax11 We have different strategy. Instead of directly cloning repo in the docker image (which is slow), we have cloned repository on the build machine. During build we update this repo with `git pull` and copy it into docker context with `COPY`.

michael-ciniawsky commented on README.md:
```diff
@@ -47,8 +47,11 @@ module.exports = {
-|**`cacheIdentifier`**|`{String}`|`cache-loader:{version} {process.env.NODE_ENV}`|Provide an invalidation identifier which is used to generate the hashes. You can use it for extra dependencies of loaders.|
+|**`cacheDirectory`**|`{String}`|`path.resolve('.cache-loader')`|Provide a cache directory where cache items should be stored (used for default read/write implementation)|
+|**`cacheIdentifier`**|`{String}`|`cache-loader:{version} {process.env.NODE_ENV}`|Provide an invalidation identifier which is used to generate the hashes. You can use it for extra dependencies of loaders.  (used for default read/write implementation)|
+|**`cacheKey`**|`{Function(options, request) -> {string}}`|`undefined`|Allows you to override default cache key generator.|
```
Please finally move this to be the first item in the options table (before `cacheDirectory`) && `-> {string} ` => `-> {String}`

michael-ciniawsky commented on src/index.js:
```diff
@@ -35,62 +45,32 @@ function loader(...args) {
-  const options = Object.assign({}, defaultOptions, loaderOptions);
-  const { cacheIdentifier, cacheDirectory } = options;
+  const options = Object.assign({}, defaults, loaderUtils.getOptions(this));
   const data = dataInput;
```
```diff
- const data = dataInput
```

michael-ciniawsky commented on src/index.js:
```diff
@@ -123,4 +103,48 @@ function digest(str) {
+        return;
+      }
+
+      directories.add(dirname);
```
`\n`

michael-ciniawsky commented on src/index.js:
```diff
@@ -123,4 +103,48 @@ function digest(str) {
+
+function cacheKey(options, request) {
+  const { cacheIdentifier, cacheDirectory } = options;
+  const hash = digest(`${cacheIdentifier}\n${request}`);
```
`\n`

michael-ciniawsky commented on src/index.js:
```diff
@@ -35,62 +45,32 @@ function loader(...args) {
-  const hash = digest(`${cacheIdentifier}\n${remainingRequest}`);
-  const cacheFile = path.join(cacheDirectory, `${hash}.json`);
+  const { read: readFn, cacheKey: cacheKeyFn } = options;
+
```
```js

const data = dataInput

data.cacheKey = cacheKeyFn(options, remainingRequest)
data.remainingRequest = remainingRequest

```

Mordred commented:
@michael-ciniawsky Changed. 

Can I ask you why are these `\n` changes so important? Current `index.js` has no newlines in function definitions and I tried to match that style :)

michael-ciniawsky commented:
> Can I ask you why are these \n changes so important? 

Of course, Code Readability. I know it is annoying and doesn't add anything, but that code needs to be maintained and whitespace in particular is one of the cheapest ways to improve readability.

>  Current index.js has no newlines in function definitions and I tried to match that style :)

Yep but it definitely should have and if no one is pesky about it the tendency is to become one big 'BLOB' again. JSDoc is also missing 😛

michael-ciniawsky commented:
Anything left TODO left here ? @sokra  @webpack-contrib/org-maintainers please review when time :)

d3viant0ne commented:
Looks like @Mordred has completed all the review comments. I don't personally see any other issues that stand out.

Unless Tobias has an objection, I'd say let it rip. :)

hulkish commented:
LGTM. Although... I know this applies more to this project as a whole - would be nice to see some tests. 

Not really practical to hold this up for that reason, though.

d3viant0ne commented:
@hulkish - @michael-ciniawsky setup a base test configuration & we are working through the repos that don't already have suites. Just haven't gotten here yet.
